csharp/tests/ParallelSelectExtTest.cs

```diff
@@ -0,0 +1,233 @@
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+using NUnit.Framework;
+
+namespace ArmoniK.Utils.Tests;
+
+public class ParallelSelectExtTest
+{
+  [Test]
+  [TestCase(-1, 0)]
+  [TestCase(-1, 1)]
+  [TestCase(-1, 4)]
+  [TestCase(-1, 1000)]
+  [TestCase(0,  0)]
+  [TestCase(0,  1)]
+  [TestCase(0,  4)]
+  [TestCase(0,  100)]
+  [TestCase(1,  0)]
+  [TestCase(1,  1)]
+  [TestCase(1,  4)]
+  [TestCase(1,  10)]
+  [TestCase(2,  0)]
+  [TestCase(2,  1)]
+  [TestCase(2,  4)]
+  [TestCase(2,  20)]
+  public async Task ParallelSelectShouldWork(int parallelism, int n)
+  {
+    var x = await GenerateInts(n)
+            .ParallelSelect(parallelism,
+                            AsyncIdentity(10))
+            .ToListAsync().ConfigureAwait(false);
+    var y = GenerateInts(n)
+      .ToList();
+    Assert.That(x, Is.EqualTo(y));
+  }
+
+  [Test]
+  [TestCase(-1, 0)]
+  [TestCase(-1, 1)]
+  [TestCase(-1, 4)]
+  [TestCase(-1, 1000)]
+  [TestCase(0, 0)]
+  [TestCase(0, 1)]
+  [TestCase(0, 4)]
+  [TestCase(0, 100)]
+  [TestCase(1, 0)]
+  [TestCase(1, 1)]
+  [TestCase(1, 4)]
+  [TestCase(1, 10)]
+  [TestCase(2, 0)]
+  [TestCase(2, 1)]
+  [TestCase(2, 4)]
+  [TestCase(2, 20)]
+  public async Task ParallelSelectAsyncShouldWork(int parallelism, int n)
+  {
+    var x = await GenerateIntsAsync(n, 10)
+                  .ParallelSelect(parallelism,
+                                  AsyncIdentity(10))
+                  .ToListAsync().ConfigureAwait(false);
+    var y = GenerateInts(n)
+      .ToList();
+    Assert.That(x, Is.EqualTo(y));
+  }
+
+  [Test]
+  [TestCase(-1,  10000)]
+  [TestCase(0,   100)]
+  [TestCase(1,   10)]
+  [TestCase(2,   20)]
+  [TestCase(10,  100)]
+  [TestCase(100, 1000)]
+  public async Task ParallelSelectLimitShouldWork(int parallelism, int n)
+  {
+    var counter    = 0;
+    var maxCounter = 0;
+    var delay      = 1000;
+
+    async Task<int> IdentityAsync(int x)
+    {
+      var count = Interlocked.Increment(ref counter);
+      var max   = maxCounter;
+      while (count > maxCounter)
+      {
+        max = Interlocked.CompareExchange(ref maxCounter,
+                                          count,
+                                          max);
+      }
+
+      await Task.Delay(delay)
+                .ConfigureAwait(false);
+      Interlocked.Decrement(ref counter);
+      return x;
+    }
+
+    var x = await GenerateInts(n)
+                  .ParallelSelect(parallelism, IdentityAsync)
+                  .ToListAsync()
+                  .ConfigureAwait(false);
+    var y = GenerateInts(n)
+      .ToList();
+    Assert.That(x, Is.EqualTo(y));
+
+    if (parallelism > 0)
+    {
+      Assert.That(maxCounter, Is.EqualTo(parallelism));
+    }
+
+    if (parallelism == 0)
+    {
+      Assert.That(maxCounter, Is.EqualTo(Environment.ProcessorCount));
+    }
+
+    if (parallelism < 0)
+    {
+      Assert.That(maxCounter, Is.EqualTo(n));
+    }
+  }
+  [Test]
```

Comment:
probably missing empty line above. We'll need to setup formatting check in this repo soon